unify split_by_effort and split_by_cost

oskooi opened the pull request:
Closes #1498.

This PR unifies `split_by_cost` and `split_by_effort` based on the outline in #1498. `split_by_effort` has been removed entirely and been made a subset of `split_by_cost` which now contains a boolean flag to specify whether to use either the fragment statistics or the pixel count as the cost function for the given `grid_volume`. The original recursive bisection to chunk subdivision is retained (which is cache oblivious as described in #1492). `split_by_cost` now works for an arbitrary number of chunks and is no longer based on prime factors. The runtime performance of `split_by_effort` is improved as it computes the entire chunk division in one go rather than the previous approach which involved separate recursive calls to determine each chunk.

A comparison of the chunk layouts for the two different methods using a 2d square cell with just vacuum is shown below. The two methods produce nearly identical results in most cases although there are small differences (which are highlighted in the figures). These differences are the result of rounding errors in the cost function using the fragment statistics. 

```py
import meep as mp

import matplotlib
matplotlib.use('agg')
import matplotlib.pyplot as plt

sim = mp.Simulation(resolution=10,
                    cell_size=mp.Vector3(10,10,0),
                    split_chunks_evenly=False)

sim.visualize_chunks()
```

![chunk_layout_2d_batch1](https://user-images.githubusercontent.com/7152530/107719010-0c267400-6c8c-11eb-8499-a4a8dd28fcb7.png)

![chunk_layout_2d_batch2](https://user-images.githubusercontent.com/7152530/107719013-0e88ce00-6c8c-11eb-8572-e036f02a4745.png)

All tests in the `make check` suite pass *except* for the subtest `test_cyl_metal_mirror_nonlinear` in `meep/tests/symmetry.cpp` which has been disabled:

https://github.com/NanoComp/meep/blob/7067536d9db2d229f7717b49f8bdb5abab0c3e10/tests/symmetry.cpp#L135-L176

This failing test involves a z-mirror symmetry plane in a cell with cylindrical coordinates which seems to be affected by the small change in the `split_by_effort` chunk division method.

```
meep: error in test_cyl_metal_mirror nonlinear vacuum
meep: error in test_cyl_metal_mirror nonlinear vacuum
imaginary part = 0 differs by 1.51084e-09 from -1.51084e-09
This gives a fractional error of 1
ez differs by 6.69695e-13-1.51084e-09i from 6.69695e-13-1.51084e-09i
This comes out to a fractional error of 1
Right now I'm looking at z = 0.5, r = 0.501, time 0.0625
```

oskooi commented:
Changing to `ttot = 10.0` from `ttot = 3.0` and `field_check_time = 8.0`.
```py
meep: error in test_cyl_metal_mirror nonlinear vacuum
meep: error in test_cyl_metal_mirror nonlinear vacuum
real part = 0.00432538 differs by -8.74538e-05 from 0.00441284
This gives a fractional error of 0.019818
er differs by 8.74538e-05-8.30032e-05i from 0.00441284-0.00354327i
This comes out to a fractional error of 0.0213051
Right now I'm looking at z = 0.5, r = 0.01, time 8
```

stevengj commented:
Since the only thing that this PR should have changed, I would print out the ivecs of the chunk coordinates in this PR vs master for the failing test, to see what changed.

oskooi commented:
For the failing test, this PR and master are producing *different* chunk divisions for the 2d cell using cylindrical coordinates as shown in the figure below. The figure shows the `ivecs` of the `litte_corner` (blue) and `big_corner` (red) for the `grid_volume` of each chunk.

![PR1499](https://user-images.githubusercontent.com/7152530/108252783-e0810f00-710d-11eb-9bc0-189b673b031d.png)

The change in the chunk division introduced by this PR is due to this line:

https://github.com/NanoComp/meep/blob/761e31b738e4e867a46d5253d2a1ea2b27dafe01/src/vec.cpp#L1101-L1102

However, it seems that the chunk division is *not* the source of the test failure because removing the nonlinearities from the test (deleting the lines with `set_chi3` or setting `chi3` internally to 0) causes it to pass:

https://github.com/NanoComp/meep/blob/761e31b738e4e867a46d5253d2a1ea2b27dafe01/tests/symmetry.cpp#L145-L146

Perhaps this suggests there is an unrelated bug somewhere in `structure_chunk::set_chi3` for cylindrical coordinates?

https://github.com/NanoComp/meep/blob/761e31b738e4e867a46d5253d2a1ea2b27dafe01/src/structure.cpp#L871-L905

oskooi commented:
Outputting the `split_measure`, `longest_axis` during the chunk splitting:

```
Testing Z mirror symmetry in Cylindrical...
Halving computational cell along direction z
Splitting into 2 chunks by voxels
split: z (direction), 80.00000 (split_measure), r (longest axis)
split: r (direction), 80.00000 (split_measure), r (longest axis)
```
```
Splitting into 2 chunks by voxels
split: z (direction), 128.00000 (split_measure), z (longest axis)
split: r (direction), 128.00000 (split_measure), z (longest axis)

```

stevengj commented on src/vec.cpp:
```diff
@@ -1094,46 +1019,20 @@ void grid_volume::find_best_split(int desired_chunks, int &best_split_point,
-        // more than 30% better than the best_split_measure. This is a heuristic to prefer lower
-        // communication costs when the split_measure is somewhat close.
-        // TODO: Use machine learning to get a cost function for the communication instead of hard
-        // coding 0.3
```
Probably we should keep this comment (and put it before the `d == longest_axis` check).

oskooi replied:
Fixed.

oskooi commented:
The failing tests described in #1506 have been disabled for the time being because they are unrelated to this PR.
